File: kthlargestmaxheap.py
# // Time Complexity :O(nlog(n-k))
# // Space Complexity :O(n-k),elements in the heap
# // Did this code successfully run on Leetcode :yes
# // Any problem you faced while coding this :no


# // Your code here along with comments explaining your approach
from heapq import heappush as insert
from heapq import heappop as remove
class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        maxheap=[]
        for i in range(len(nums)):
            insert(maxheap,(-nums[i]))
            # print(maxheap)
            if (len(maxheap)-1)>(len(nums)-k):
                remove(maxheap)
        # print(maxheap)
            
        return -(maxheap[0])
        
        

File: KthLargestMaxHeap.java
class Solution {
    //min heap
    //tc- o(nlog(n-k)) where is k is kth largest element
    //sc- o(n-k)
    public int findKthLargest(int[] nums, int k) {
        PriorityQueue<Integer> pq = new PriorityQueue<>((a,b)->(b-a));
        int min = Integer.MAX_VALUE;
        for(int num : nums)
        {
            pq.add(num);
            
            if(pq.size()>nums.length-k)
            {
                
                min = Math.min(min,pq.poll());
               
            }
        }
        return min;

    }
}